descartes_trajectory/src/cartesian_interpolator.cpp

```diff
@@ -0,0 +1,199 @@
+  std::vector<double> seed(robot_model_->getDOF(),0.0f);
+  for(unsigned int i = 1; i < coarse_traj.size();i++)
+  {
+    const CartTrajectoryPt& c1 = coarse_traj[i-1];
+    const CartTrajectoryPt& c2 = coarse_traj[i];
+    if(c1.getNominalCartPose(seed,*robot_model_,p_start) &&
+        c2.getNominalCartPose(seed,*robot_model_,p_end))
+    {
+
+      Eigen::Vector3d p1 = p_start.translation();
+      Eigen::Vector3d p2 = p_end.translation();
+      double norm = (p2-p1).norm();
+
+      ROS_DEBUG_STREAM("Distance for segment "<<i-1<<": "<<(p2-p1).norm());
+      double dt = norm/tool_speed_;
+      ft[i-1] = dt;
+    }
+    else
+    {
+      ROS_ERROR_STREAM("Failed to get nominal cartesian poses while calculating time for segment "
+          <<c1.getID()<<" and "<<c2.getID());
+      return false;
+    }
+  }
+
+  return true;
+}
+
+bool CartesianInterpolator::initialize(descartes_core::RobotModelConstPtr robot_model,
+                                       double tool_speed,double interpolation_interval, double zone_radius)
+{
+  robot_model_ = robot_model;
+  tool_speed_ = tool_speed;
+  interpolation_interval_= interpolation_interval;
+  zone_radius_ = zone_radius;
+
+  return true;
+}
+
+bool CartesianInterpolator::interpolate(const std::vector<descartes_trajectory::CartTrajectoryPt>& coarse_traj,
+                 std::vector<descartes_core::TrajectoryPtPtr>& interpolated_traj)
+{
+
+  std::vector<double> segment_times;
+  if(!computeFullSegmentTimes(coarse_traj,segment_times))
+  {
+    ROS_ERROR_STREAM("Failed to compute trajectory segment time values");
+    return false;
+  }
+
+  // estimating number of total points in interpolated trajectory
+  std::size_t total_points = 0;
+  for(std::size_t i = 0; i < segment_times.size();i++)
+  {
+    total_points += std::ceil(segment_times[i]/interpolation_interval_) + 1;
+  }
+  interpolated_traj.clear();
+  interpolated_traj.reserve(total_points);
+
+  // interpolating segments
+
+   tf::Transform p_start, p_end;
+  Eigen::Affine3d eigen_start, eigen_end;
+  std::vector<double> seed(robot_model_->getDOF(),0.0f);
+  std::vector<TrajectoryPtPtr> segment_points;
+  for(std::size_t i = 0; i < segment_times.size();i++)
+  {
+    const CartTrajectoryPt& ps = coarse_traj[i];
+    const CartTrajectoryPt& pe = coarse_traj[i+1];
+    ps.getNominalCartPose(seed,*robot_model_,eigen_start);
+    pe.getNominalCartPose(seed,*robot_model_,eigen_end);
+
+    tf::poseEigenToTF(eigen_start,p_start);
+    tf::poseEigenToTF(eigen_end,p_end);
+    LinearSegment lsegment = LinearSegment(p_start,p_end,segment_times[i],interpolation_interval_);
+    if(lsegment.interpolate(segment_points))
+    {
+      interpolated_traj.push_back(TrajectoryPtPtr(new CartTrajectoryPt(ps)));
+      interpolated_traj.insert(interpolated_traj.end(), segment_points.begin(),segment_points.end());
+
+      if(i == segment_times.size() - 1)
+      {
+        interpolated_traj.push_back(TrajectoryPtPtr(new CartTrajectoryPt(ps)));
```

Comment:
I'm a little confused on what's going on here. On the last iteration you add the start point of the last interpolation to the end? Should this be the variable "pe" instead?